pkg/armhelpers/interfaces.go

```diff
@@ -95,6 +95,12 @@ type AKSEngineClient interface {
 	// DeployTemplate can deploy a template into Azure ARM
 	DeployTemplate(ctx context.Context, resourceGroup, name string, template, parameters map[string]interface{}) (resources.DeploymentExtended, error)
 
+	// ListVirtualMachineImages return a list of images
```

Comment:
What do you think @devigned? Should we create a `pkg/armhelpers/v2/interfaces.go` to create the additive method changes here? Making these changes in place will indeed break contracts of pre-existing `AKSEngineClient` type usages.

Reply:
I agree with you. This should be moved into a new interface. Perhaps, there should be an interface like below:

```go
type VMImageFetcher interface {
	ListVirtualMachineImages(ctx context.Context, top int32, location, publisherName, offer, skus, filter, orderBy string) (compute.ListVirtualMachineImageResource, error)
 	GetVirtualMachineImage(ctx context.Context, location, publisherName, offer, skus, version string) (compute.VirtualMachineImage, error)
}
```

If the client implements that type, then the image checking functionality could be used. For example:

```go
if imageFetcher, ok := akseClient.(VMImageFetcher); ok {
  // do the image checking stuff...
}
```

That doesn't limit the ability of AKS Engine to provide additional functionality and it allows that new functionality to be non-breaking to existing consumers. I favor this method since it is a "nice to have" behavior, not something that is required to function. That is to say, the client will still build a cluster without the ability to check if an image exists, but there may be an error raised by ARM at a later time stating the image does not exist. It is not optimal behavior since with the ability to check if an image exists prior to the deployment, we could respond quicker. In either case, the behavior will reach the same end state.